File: code/ch13/ruby/norm_code.rb
# Exercise 11 (§13.6.2) — the inherited, undocumented norm_code, Ruby variant.
# nil plays Python's None; ArgumentError plays ValueError; the [[:alnum:]] regex
# plays str.isalnum (nonempty, all letters/digits). Ruby's "" is truthy, so the
# final `s or ...` needs an explicit empty? check.
# Runs standalone: ruby norm_code.rb
def norm_code(s, strict: false)
  return strict ? nil : "" if s.nil?
  s = s.strip.upcase.delete("-")
  s = s[0, 8] if s.length > 8
  raise ArgumentError, s if strict && s !~ /\A[[:alnum:]]+\z/
  s.empty? ? (strict ? nil : "") : s
end

# characterization pins, matching the Python original's observed behavior
raise unless norm_code(nil) == ""
raise unless norm_code(nil, strict: true).nil?
raise unless norm_code(" ab-12 ") == "AB12"
raise unless norm_code("abcdefghij") == "ABCDEFGH" # truncated to 8
raise unless norm_code("a!b") == "A!B"             # lenient keeps the junk
begin
  norm_code("a!b", strict: true)
  raise "expected ArgumentError"
rescue ArgumentError => e
  raise unless e.message == "A!B"
end
begin
  norm_code("", strict: true)                      # empty fails isalnum too
  raise "expected ArgumentError"
rescue ArgumentError
  nil
end
raise unless norm_code("  ") == ""
raise unless norm_code("--------") == ""
puts "all characterization pins hold"

File: code/ch13/java/NormCode.java
// Exercise 11 (§13.6.2) — the inherited, undocumented norm_code, Java variant.
// null plays Python's None; IllegalArgumentException plays ValueError; the
// allMatch check plays str.isalnum (nonempty, all letters/digits).
// Run: java -ea NormCode.java
public class NormCode {
  static String normCode(String s, boolean strict) {
    if (s == null) return strict ? null : "";
    s = s.strip().toUpperCase().replace("-", "");
    if (s.length() > 8) s = s.substring(0, 8);
    boolean alnum = !s.isEmpty() && s.chars().allMatch(Character::isLetterOrDigit);
    if (strict && !alnum) throw new IllegalArgumentException(s);
    return s.isEmpty() ? (strict ? null : "") : s;
  }

  public static void main(String[] args) {
    // characterization pins, matching the Python original's observed behavior
    assert normCode(null, false).equals("");
    assert normCode(null, true) == null;
    assert normCode(" ab-12 ", false).equals("AB12");
    assert normCode("abcdefghij", false).equals("ABCDEFGH"); // truncated to 8
    assert normCode("a!b", false).equals("A!B");             // lenient keeps the junk
    try {
      normCode("a!b", true);
      throw new AssertionError("expected IllegalArgumentException");
    } catch (IllegalArgumentException e) {
      assert e.getMessage().equals("A!B");
    }
    try {
      normCode("", true);                                    // empty fails isalnum too
      throw new AssertionError("expected IllegalArgumentException");
    } catch (IllegalArgumentException e) {
      assert e.getMessage().equals("");
    }
    assert normCode("  ", false).equals("");
    assert normCode("--------", false).equals("");
    System.out.println("all characterization pins hold");
  }
}
